libraries/AP_Torqeedo/AP_Torqeedo_TQBus.cpp

```diff
@@ -0,0 +1,1119 @@
+
+    // @Param: SLEW_TIME
+    // @DisplayName: Torqeedo Throttle Slew Time
+    // @Description: Torqeedo slew rate specified as the minimum number of seconds required to increase the throttle from 0 to 100%.  Higher values cause a slower response, lower values cause a faster response.  A value of zero disables the limit
+    // @Units: s
+    // @Range: 0 5
+    // @Increment: 0.1
+    // @User: Advanced
+    AP_GROUPINFO("SLEW_TIME", 6, AP_Torqeedo_TQBus, _slew_time, 2.0),
+
+    // @Param: DIR_DELAY
+    // @DisplayName: Torqeedo Direction Change Delay
+    // @Description: Torqeedo direction change delay.  Output will remain at zero for this many seconds when transitioning between forward and backwards rotation
+    // @Units: s
+    // @Range: 0 5
+    // @Increment: 0.1
+    // @User: Advanced
+    AP_GROUPINFO("DIR_DELAY", 7, AP_Torqeedo_TQBus, _dir_delay, 1.0),
+
+    // @Param: SRV_FUNC
+    // @DisplayName: Torqeedo motor servo function
+    // @Description: Torqeedo motor servo function. Default is kthrottle
+    AP_GROUPINFO("SRV_FUNC", 8, AP_Torqeedo_TQBus, _channel_function, 70),
+
+    AP_GROUPEND
+};
+
+AP_Torqeedo_TQBus::AP_Torqeedo_TQBus(AP_Torqeedo &_frontend,
+                                     AP_Torqeedo::Torqeedo_State &_state,
+                                     AP_Torqeedo_Params& _params,
+                                     uint8_t serial_instance) :
+    AP_Torqeedo_Backend_Serial(_frontend, _state, _params, serial_instance)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    state.var_info = var_info;
+}
+
+// initialise driver
+void AP_Torqeedo_TQBus::init(void)
+{
+    // exit immediately if not enabled
+    if (!enabled()) {
+        return;
+    }
+
+    // only init once
+    // Note: a race condition exists here if init is called multiple times quickly before thread_main has a chance to set _initialise
+    if (_initialised) {
+        return;
+    }
+
+    // setup a name for the thread
+    char name[32];
+    snprintf(name, sizeof(name), "torqeedo-TQ-%u", (uint8_t)(size_t)&_serial_instance);
+
+    //create background thread to process serial input and output
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_Torqeedo_TQBus::thread_main, void), name, 2048, AP_HAL::Scheduler::PRIORITY_RCOUT, 1)) {
+        return;
+    }
+}
+
+
+// update driver state
+void AP_Torqeedo_TQBus::update()
+{
+    if (!healthy()) {
+        set_status(AP_Torqeedo::Status::NoData);
+    } else {
+        set_status(AP_Torqeedo::Status::Good);
+    }
+}
+
+// initialise serial port and gpio pins (run from background thread)
+bool AP_Torqeedo_TQBus::init_internals()
+{
+    _uart->begin(TORQEEDO_SERIAL_BAUD);
+    _uart->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+    _uart->set_unbuffered_writes(true);
+
+    // if using tiller connection set on/off pin for 0.5 sec to turn on battery
+    if (_type == ConnectionType::TYPE_TILLER) {
```

Comment:
Ah, I see now why there are duplicate Type parameters.  I think we could use the same TRQBus backend but have two types.  One for tiller and one for Motor.  The frontend TYPE parameter could have two values (1:TRQ Tiller, 2:TQR Motor) and which one is selected could be passed into the constuctor.